### PR TITLE
Add procedure calling convention.

### PR DESCRIPTION
**Pull request by Calring**
docs/LoongArch-ELF-ABI-EN.adoc

### CODE REVIEW

**Comment by xen0n** [comment]
Why are you re-creating PRs every time you make a revision? Reviewers can't easily see whether their comments have been addressed. You may just force-push to your PR branch, no need to delete or do anything else.

**Comment by Calring** [comment]
>

**Comment by Calring** [comment]
> Why are you re-creating PRs every time you make a revision? Reviewers can't easily see whether their comments have been addressed. You may just force-push to your PR branch, no need to delete or do anything else.

OK.Got it.

**Comment by yetist** [comment]
这个是之前 #26 和 #30 的讨论吧？ 之前的讨论历史都跟丢了。

一个PR，多次修改的话，直接强推就好，这样所有的讨论历史都在。就是说我们可以在 #26 上一直完善到最终状态，并不需要新开 #30 和 #32 来重新讨论。

**Comment by xen0n** [comment]
> 这个是之前 #26 和 #30 的讨论吧？ 之前的讨论历史都跟丢了。
> 
> 一个PR，多次修改的话，直接强推就好，这样所有的讨论历史都在。就是说我们可以在 #26 上一直完善到最终状态，并不需要新开 #30 和 #32 来重新讨论。

按照贴主之前的操作方式（删分支再重推），在 GitHub 上是不可能关联到先前的 PR 了（在删除分支时就不可逆地关掉了）。贴主需要做的是熟悉 GitHub 上较为流行的协作流程。幸好前面讨论的内容（要求 rebase，修正 `s0-s8` 没包括 `s9` 的写法）都比较小，而且改掉了。

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 1]
```diff
@@ -628,3 +628,140 @@ with check 32-bit unsigned overflow
...
+. GRLEN < WOA ≤ 2 × GRLEN
+.. Only fixed-point members.
+The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR. If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+
+.. Only float-point members.
+i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.
+The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR. If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+
+.. Both fixed-point and floating-point members.
+... The structure has one double member, and
```
Comment (sterling-teng) [inline docs/LoongArch-ELF-ABI-EN.adoc:706]: and?
It looks like the Enter key was pressed by mistake in some advanced editor.  :)

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 2]
```diff
@@ -628,3 +628,140 @@ with check 32-bit unsigned overflow
...
+.... Multiple fixed-point members
+The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR. If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+... The structure has float and fixed-point member
+The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR. If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+
+. WOA > 2 × GRLEN
+
+.. It’s passed by reference and are replaced in the argument list with the address.If there is an available GAR, the reference is passed in the GAR, and passed on the stack if no GAR is available.
```
Comment (sterling-teng) [inline docs/LoongArch-ELF-ABI-EN.adoc:716]: the address.If there

A space is missing after the period.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 3]
```diff
@@ -628,3 +628,140 @@ with check 32-bit unsigned overflow
...
+
+Structure and scalars passed on the stack are aligned to the greater of the type alignment and GRLEN bits, but never more than the stack alignment.
+
+==== Union
```
Comment (sterling-teng) [inline docs/LoongArch-ELF-ABI-EN.adoc:720]: Miss a blank line.

There are many more below, please standardize the format.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 4]
```diff
@@ -628,3 +628,140 @@ with check 32-bit unsigned overflow
...
+Union is passed in GAR or stack;
```
Comment (sterling-teng) [inline docs/LoongArch-ELF-ABI-EN.adoc:721]: The end of the sentence probably needs a colon, not a semicolon.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 5]
```diff
@@ -628,3 +628,140 @@ with check 32-bit unsigned overflow
...
+
+.. The variadic arguments are passed in a GAR, or on the stack by value if no GAR is available.
+
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. The variadic arguments are passed in an aligned GAR pair (i.e., the first register in the pair is even-numbered), or on the stack by value if none is available.
+
+. WOA > 2 × GRLEN
+
+.. It’s passed by reference and are replaced in the argument list with the address.If there is an available GAR, the reference is passed in the GAR, and passed on the stack if no GAR is available
```
Comment (sterling-teng) [inline docs/LoongArch-ELF-ABI-EN.adoc:751]: no GAR is available.

**Review by sterling-teng — COMMENTED** [review]
We do not like a paragraph as a line because it is not easy to review. Normally, we break the line after a period, like this:

xxx, xxxxxxx.
xxxxx,xxxxxxx.

xxxxxxx,xxxxxxxx.

xxxxxx,xxxxxxxxxxxxxx.
xxxxxxx,xxxxxxx,xxxxxxxx.

The doc is generally fine. but could it be better with few more polish work?

**Review by sterling-teng — APPROVED** [review]

**Comment by Calring** [comment]
> We do not like a paragraph as a line because it is not easy to review. Normally, we break the line after a period, like this:
> 
> xxx, xxxxxxx. xxxxx,xxxxxxx.
> 
> xxxxxxx,xxxxxxxx.
> 
> xxxxxx,xxxxxxxxxxxxxx. xxxxxxx,xxxxxxx,xxxxxxxx.
> 
> The doc is generally fine. but could it be better with few more polish work?

Great! I've adjusted them accordingly to your suggestions. If there are still mistakes, please continue to put forward.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 6]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+== Procedure Calling Convention
+
+The LoongArch procedure calling convention passes arguments in registers as much as possible (i.e. Floating-point arguments are passed in floating-point registers and non floating-point arguments are passed in general-purpose registers as much as possible)，and then considers passing arguments on the stack. This is a basic principle of LoongArch procedure calling convention.
+
+The argument registers that passes arguments are:
+
+. Eight floating-point registers `fa0-fa7`, and `fa0-fa1` are also used to return values. They are all used to pass floating-point arguments.
+
+. Eight general-purpose registers `a0-a7`, and `a0-a1` are also used to return values. Generally, they are used to pass fixed-point arguments and floating-point arguments when the floating-point register is not available. And some other uses will be described below.Argument bit fields are stored in little endian mode. In addition, the subroutine should ensure that the values of general-purpose registers `s0-s9` and floating-point registers `fs0-fs7` are preserved across procedure calls.
```
Comment (ChenghuaXu) [inline docs/LoongArch-ELF-ABI-EN.adoc:640]: Miss a space before 'Argument bit ...'.

**Review by ChenghuaXu — CHANGES_REQUESTED** [review]

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 6 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc]: Ok

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 7]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
 |Linux, Glibc
 |/lib32/ld-linux-loongarch-ilp32s.so.1
 |===
+
+== Procedure Calling Convention
+
+The LoongArch procedure calling convention passes arguments in registers as much as possible (i.e. Floating-point arguments are passed in floating-point registers and non floating-point arguments are passed in general-purpose registers as much as possible)，and then considers passing arguments on the stack. This is a basic principle of LoongArch procedure calling convention.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:634]: 看起来“and then”的一小句是中式英语：“之后考虑在栈上传参”，接下来“这是 LoongArch 过程调用约定的基本原则”这一句分开说，感觉也是中式英语。另外有一个全角逗号。

修改一下措辞：

```suggestion
The basic principle of the LoongArch procedure calling convention is to pass arguments in registers as much as possible (i.e. floating-point arguments are passed in floating-point registers and non floating-point arguments are passed in general-purpose registers, as much as possible); arguments are passed on the stack only when no appropriate register is available.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 8]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+The argument registers that passes arguments are:
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:636]: 冗余

```suggestion
The argument registers are:
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 9]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+. Eight floating-point registers `fa0-fa7`, and `fa0-fa1` are also used to return values. They are all used to pass floating-point arguments.
+
+. Eight general-purpose registers `a0-a7`, and `a0-a1` are also used to return values. Generally, they are used to pass fixed-point arguments and floating-point arguments when the floating-point register is not available. And some other uses will be described below. Argument bit fields are stored in little endian mode. In addition, the subroutine should ensure that the values of general-purpose registers `s0-s9` and floating-point registers `fs0-fs7` are preserved across procedure calls.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:640]: - 加一个逗号，限定清楚 when 子句的修饰范围
- “……以及浮点参数，当浮点寄存器不可用”——“not available”是指浮点寄存器在硬件上存在但用完了，还是硬件未实现（软浮点 ABI 之类情况）？
- 其他一些措辞修改

```suggestion
. Eight general-purpose registers `a0-a7`, with `a0-a1` reused to return values.

Generally, the GARs are used to pass fixed-point arguments, and floating-point arguments when the floating-point register is not available.
Bit fields are stored in little endian.
In addition, subroutines should ensure that the values of general-purpose registers `s0-s9` and floating-point registers `fs0-fs7` are preserved across procedure calls.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 10]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+==== Note
+
+In this document, *GRLEN* is the bit width of general-purpose register, *FRLEN* is the bit width of float-point register and *WOA* is the bit width of the argument. The argument registers are collectively denoted as *AR* (*GAR* or *FAR*), where the general-purpose argument register is denoted as *GAR* and the floating-point argument register is denoted as *FAR*.
+
+=== ABI LP64D
+
+that is, *GRLEN* = 64, *FRLEN* = 64.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:648]:
```suggestion
That is, *GRLEN* = 64, *FRLEN* = 64.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 11]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+=== C Datatypes and Alignment
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:650]: `datatype` 不是一个单词，需要加空格。

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 12]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+The C datatypes and Alignment of *LP64D* refer to LP64 Data Model in the table 3.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:652]: Where's table 3?

由于这个句子整个都是中式英语，请重写一下，类似于“The C data types and alignment in the LP64D ABI are defined in <插入正确格式的链接>”这样。

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 13]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+The C types `*unsigned char*` is 8-bit unsigned integers and zero-extended when stored in a LoongArch general-purpose register. `*unsigned short*` is a 16-bit unsigned integer and is zero-extended when stored in a LoongArch general-purpose register. `*char*` and `*signed char*` are an 8-bit signed integer and sign-extended when stored in a LoongArch general-purpose register. `*short*` is a 16-bit signed integer and is sign-extended when stored in a register.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:654]: 这段要写这么多吗？感觉十分冗余的样子，写一下“无符号的基本整数类型在寄存器中均以零扩展方式存储，有符号的基本整数类型在寄存器中均以符号扩展方式存储”就好了。

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 14]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+In LoongArch *LP64D*, 32-bit types, such as `*int*`, are stored in general-purpose registers as proper sign extensions of their 32-bit values. This restriction holds even for unsigned 32-bit types.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:656]: 不用到处强调 LoongArch？

```suggestion
In the LP64D ABI, 32-bit types, such as `*int*`, are stored in general-purpose registers as proper sign extensions of their 32-bit values. This restriction holds even for unsigned 32-bit types.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 15]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+=== Argument passing
+
+Generally speaking, FARs are only used to pass floating-point arguments, GARs pass non floating-point arguments and floating-point arguments when the floating-point register is not available(long double is also passed in a pair of GARs) and the reference, arguments are passed on the stack if no AR is available.
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:660]: “available”的问题前面说过了，另外这个段落也有中式英语问题，改一下

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 16]
```diff
@@ -628,3 +628,158 @@ with check 32-bit unsigned overflow
...
+
+Arguments passed by reference may be modified by the callee.
+
+==== Scalar (that is, integer, floating-point or pointer types)
```
Comment (xen0n) [inline docs/LoongArch-ELF-ABI-EN.adoc:664]: 读者是专业人士，不用 that is 吧？一定要加的话也是建议写到这个 section 的正文开头。

Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc]: 已做修改.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 10 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc]: 已做对应修改.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 17]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+. Eight general-purpose registers `a0-a7`, with `a0-a1` reused to return values.
+
+Generally, the GARs are used to pass fixed-point arguments, and floating-point arguments when no FAR is available.
+Bit fields are stored in little endian.
+In addition, subroutines should ensure that the values of general-purpose registers `s0-s9` and floating-point registers `fs0-fs7` are preserved across procedure calls.
+
+==== Note
+
+In this document, *GRLEN* is the bit width of general-purpose register, *FRLEN* is the bit width of float-point register and *WOA* is the bit width of the argument.
+The argument registers are collectively denoted as *AR* (*GAR* or *FAR*), where the general-purpose argument register is denoted as *GAR* and the floating-point argument register is denoted as *FAR*.
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:649]: Does using *AR* abbreviation justified? It is used in just one place and does not save too much typing, but clearly hide meaning, as all abbreviations. If you look for example at RISC-V document https://riscv.org/wp-content/uploads/2015/01/riscv-calling.pdf they are not abbreviate "integrer registers" and "floating-point registers" . Same for other ABI documents 
- https://github.com/hjl-tools/x86-psABI/wiki/x86-64-psABI-1.0.pdf
- https://github.com/ARM-software/abi-aa/
and others. 
So other question, does *GAR* and *FAR* abbreviation is justified?

**Review by kant2002 — COMMENTED** [review]
Left comments about pieces which puzzle me during reading.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 18]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+
+.. Argument is passed in a single AR, or on the stack by value if none is available.
+
+... If the argument is float-point type, the argument is passed in FAR.
+if no FAR is available, it’s passed in GAR. If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, floating-point types narrower than GRLEN bits are widened to GRLEN bits, with the upper bits undefined.
+
+... If the argument is integer or pointer type, the argument is passed in GAR.
+If no GAR is available, it’s passed on the stack.
+When passed in registers or on the stack, integer scalars narrower than GRLEN bits are sign-extended to GRLEN bits.
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:684]: Sign-extension typically refers to signed integers. Are unsigned integers zero extended, or treated sign-extended as if the integer was signed?

For example, is `uint8_t` value of `0xFF` passed in as `0x00000000000000FF` or as `0xFFFFFFFFFFFFFFFF` ?

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 19]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+If exactly one register is available, the low-order GRLEN bits are passed in the register and the high-order GRLEN bits are passed on the stack.
+If no GAR is available, it’s passed on the stack.
+
+. WOA > 2 × GRLEN
+
+.. Scalars are passed by reference and are replaced in the argument list with the address.
+
+==== Structure
+
+Empty structures are ignored by C compilers which support them as a non-standard extension(same as union arguments and return values).
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:698]: Is this trying to say that empty structures do not occupy any registers or space on stack?

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 20]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.
+And the layout of the structure on the stack is consistent with that in memory.
+
+. 0 < WOA ≤ GRLEN
+
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:705]: This means that integers and trivial struct wrappers of integers are not passed the same way. Integers are sign-extented; trivial struct wrappers of integers are not sign-extended (upper bits are undefined).

Is this example correct?
```
void A(int x); // x is passed in `a0` register, upper 32-bit of `a0` are sign-extended

struct S
{
    int f;
};

void B(struct S s); // s is passed in `a0` registers, upper 32-bit of `a0` registers are undefined
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 21]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. Only fixed-point members.
+
+... The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR.
+If the number of available GAR is 1, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if there is no available GAR.
+
+.. Only float-point members.
+
+... i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:733]: Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 22]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
 |/lib32/ld-linux-loongarch-ilp32s.so.1
 |===
+
+== Procedure Calling Convention
+
+The basic principle of the LoongArch procedure calling convention is to pass arguments in registers as much as possible (i.e. floating-point arguments are passed in floating-point registers and non floating-point arguments are passed in general-purpose registers, as much as possible); arguments are passed on the stack only when no appropriate register is available.
+
+The argument registers are:
+
+. Eight floating-point registers `fa0-fa7`, and `fa0-fa1` are also used to return values. They are all used to pass floating-point arguments.
```
[the remark this answers is not in the document]
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:638]: > It was hard to parse that `fa0-fa7` used for passing arguments.

Thanks for your review and advices!

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 18 above)
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:684]: > Sign-extension typically refers to signed integers. Are unsigned integers zero extended, or treated sign-extended as if the integer was signed?
> 
> For example, is `uint8_t` value of `0xFF` passed in as `0x00000000000000FF` or as `0xFFFFFFFFFFFFFFFF` ?

both signed and unsigned are passed as `0xFFFFFFFFFFFFFFFF` .

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 19 above)
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:698]: yes, I think so.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 20 above)
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:705]: > This means that integers and trivial struct wrappers of integers are not passed the same way. Integers are sign-extented; trivial struct wrappers of integers are not sign-extended (upper bits are undefined).
> 
> Is this example correct?
> 
> ```
> void A(int x); // x is passed in `a0` register, upper 32-bit of `a0` are sign-extended
> 
> struct S
> {
>     int f;
> };
> 
> void B(struct S s); // s is passed in `a0` registers, upper 32-bit of `a0` registers are undefined
> ```

There is only one field, the upper 32-bits should be sign extened.

If  `struct S {char  a; cha b;}` , may be just passed by value, the upper 32-bits are undefined.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 21 above)
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:733]: > Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Yes, the struct is passed by two integer register(GARs).

By the way, I had asked the author that `... i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.`
He wanted to express that:
struct {long double a;} or   struct {double a; double b;}  or struct {double a; float b; float c;} or struct {float a; float b; float c; float d;}.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 23]
```diff
@@ -628,3 +628,197 @@ with check 32-bit unsigned overflow
...
+
+. Eight general-purpose registers `a0-a7`, with `a0-a1` reused to return values.
```
[the remark this answers is not in the document]
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:640]: I had told the author.
He will update it later.
Thanks ~

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 18 above)
Comment (jkotas) [reply docs/LoongArch-ELF-ABI-EN.adoc:684]: Is there a rationale behind this choice? I think that this design requires extra instructions to properly sign-extend and zero-extend the upper bits of unsigned integers. It just adds an extra overhead.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 20 above)
Comment (jkotas) [reply docs/LoongArch-ELF-ABI-EN.adoc:705]: This sounds contrary to https://github.com/loongson/LoongArch-Documentation/pull/32/files#diff-711b3e7b6a005b492898ac6d93f2d8d37c00e0831e210993a6f9dbb26c043717R699 : Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 21 above)
Comment (jkotas) [reply docs/LoongArch-ELF-ABI-EN.adoc:733]: So the only case of GRLEN < WOA ≤ 2 × GRLEN structures that is candidate for FAR is one double member and one fixed-point member. All of other cases are passed in GAR. I think that the description can be simplified and just mention this one case as an exception.

Is there a rationale for why the one double member and one fixed-point member structures are special cased like this? It sounds like a lot of complexity for very little gain.

Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:733]: > So the only case of GRLEN < WOA ≤ 2 × GRLEN structures that is candidate for FAR is one double member and one fixed-point member. All of other cases are passed in GAR. I think that the description can be simplified and just mention this one case as an exception.

Yes, there is room for improment. 

> Is there a rationale for why the one double member and one fixed-point member structures are special cased like this? It sounds like a lot of complexity for very little gain.

Now it's too late for amendment of the ABI. (By the way, the RISC-V has the same situation)
This ABI for gcc/llvm might be unconspicuous, but for JIT is a lot fo comlexity and maybe worse for gain.

For dotnet-runtime, I have to implement the ABI following this spec only.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 18 above)
Comment (shushanhf) [reply docs/LoongArch-ELF-ABI-EN.adoc:684]: > Is there a rationale behind this choice? 

This depends on the processor's architecture. Liking the LoongArch64, MIPS64, the most operations of 32-bits's instruction, liking compare, add, multiply.... , have striction on the high-32bits.


> I think that this design requires extra instructions to properly sign-extend and zero-extend the upper bits of unsigned integers. It just adds an extra overhead.

Comment (jkotas) [reply docs/LoongArch-ELF-ABI-EN.adoc]: I understand that it may be beneficial to widen arguments or return values given the processor architecture. I have seen it done in other ABIs.

However, it is very surprising to me to widen unsigned values via sign-extension as if they were signed. The other ABIs widen the unsigned values via zero-extensions. For example, RISC-V that you took a lot of inspiration from widens unsigned values via zero-extension as far as I can tell.

Comment (xen0n) [reply docs/LoongArch-ELF-ABI-EN.adoc]: Sorry for jumping in, but I think maybe @shushanhf should confirm with their teammates the actual implemented behavior in the gcc port. If the implementation is primarily derived from the RISC-V port, then such behavior is unlikely to be delibrately changed.

**Review by chenglulu326 — APPROVED** [review]

**Review by scylaac — APPROVED** [review]

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 21 above)
Comment (chenglulu326) [reply docs/LoongArch-ELF-ABI-EN.adoc]: > Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Actually, a structure with two 'double' members passed in pair of available FARS.

Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:760]: > Does this also apply to a structure that has two `double` members? Is structure with two `double` members passed in GARs?

Sorry for missing this. I've added it in new submit.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 24]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+Empty structures are ignored by C compilers which support them as a non-standard extension(same as union arguments and return values).
+Bits unused due to padding, and bits past the end of a structure whose size in bits is not divisible by GRLEN, are undefined.
+And the layout of the structure on the stack is consistent with that in memory.
+
+. 0 < WOA ≤ GRLEN
+
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
+
+.. The structure has only float-point members:
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:734]:
```suggestion
.. The structure has only floating-point members:
```
Fix up all occurrences in the doc.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 25]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+
+... One double member.
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:736]:
```suggestion
... One floating-point member.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 26]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+The argument is passed in a FAR; If no FAR is available, the value is passed in a GAR; if no GAR is available, the value is passed on the stack.
+
+... Two float members.
```
Comment (jkotas) [inline docs/LoongArch-ELF-ABI-EN.adoc:739]:
```suggestion
... Two floating-point members.
```

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 24 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:734]: Ok.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 26 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc]: I've fixed up. Thanks~

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 27]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+
+Arguments passed by reference may be modified by the callee.
+
+==== Scalar
+
+There are three cases:
+
+. 0 < WOA ≤ GRLEN
+
+.. Argument is passed in a single AR, or on the stack by value if none is available.
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:703]:
```suggestion
.. Argument is passed in a single argument register, or on the stack by value if none is available.
```

This is only place where *AR* as abbreviation used. Also expanded version like "argument register" and "argument registers" used a bit in the document too. I would advocate for removal at least this (*AR*) redundant abreviation.

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 28]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+.. A place to store the subroutine’s return address.
+
+.. A place to store the values of saved registers.
+
+.. A place for local data storage.
+
+. The stack grows downwards (towards lower addresses) and the stack pointer shall be aligned to a 128-bit boundary upon procedure entry.
+The first argument passed on the stack is located at offset zero of the stack pointer on function entry; following arguments are stored at correspondingly higher addresses.
+
+. Procedures must not rely upon the persistence of stack-allocated data whose addresses lie below the stack pointer.
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:847]:
```suggestion
. Procedures must not rely upon the persistence of stack-allocated data whose addresses lies below the stack pointer.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 29]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+.. The structure has only fixed-point members.
+If there is an available GAR, the structure is passed through the GAR by value passing; If no GAR is available, it’s passed on the stack.
+
+.. The structure has only floating-point members:
+
+... One floating-point member.
+The argument is passed in a FAR; If no FAR is available, the value is passed in a GAR; if no GAR is available, the value is passed on the stack.
+
+... Two floating-point members.
+The argument is passed in a pair of available FAR, with the low-order float member bits in the lower-numbered FAR and the high-order float member bits in the higher-numbered FAR.
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:740]:
```suggestion
The argument is passed in a pair of available FAR, with the low-order floating-point member bits in the lower-numbered FAR and the high-order floating-point member bits in the higher-numbered FAR.
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 30]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+If the number of available FAR is less than 2, it’s passed in a GAR, and passed on the stack if no GAR is available.
+
+.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:743]:
```suggestion
.. The structure has both fixed-point and floating-point members, i.e. the structure has one floating-point member and
```

File: docs/LoongArch-ELF-ABI-EN.adoc [hunk 31]
```diff
@@ -657,3 +657,191 @@ with check 32-bit unsigned overflow
...
+. GRLEN < WOA ≤ 2 × GRLEN
+
+.. Only fixed-point members.
+
+... The argument is passed in a pair of available GAR, with the low-order bits in the lower-numbered GAR and the high-order bits in the higher-numbered GAR.
+If only one register is available, the low-order bits are in the GAR and the high-order bits are on the stack, and passed on the stack if no GAR is available.
+
+.. Only floating-point members.
+
+... i.e. The structure has one long double member or one double member and two adjacent float members or 3-4 float members.
```
Comment (kant2002) [inline docs/LoongArch-ELF-ABI-EN.adoc:760]:
```suggestion
... i.e. The structure has one long double member or one double member and two adjacent floating-point members or 3-4 floating-point members.
```

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 27 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:703]: Great.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 30 above)
Comment (kant2002) [reply docs/LoongArch-ELF-ABI-EN.adoc:743]: Other then that. When read this line, I initially though that it is half-written. Then I realize that this is trick to simplify shared naming in headers.
I would suggest add indication that this is not full header, for example like this
```
.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and...

... multiple fixed-point members.

... only one fixed-point member.
```
or other way
```
.. The structure has both fixed-point and floating-point members, i.e. the structure has one float member and

... ...multiple fixed-point members.

... ...only one fixed-point member.
```

`...` serve as visual hint that this is header would be not complete and some continuation lays somewhere down the road.

Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:743]: Thanks a lot.

File: docs/LoongArch-ELF-ABI-EN.adoc (hunk 31 above)
Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:760]: Actually, I want to express data type `float` here. As I kown, floatting-point include data type `float` and `double`... Is my understanding right?

Comment (xen0n) [reply docs/LoongArch-ELF-ABI-EN.adoc:760]: > Actually, I want to express data type `float` here. As I kown, floatting-point refers to data type `float` and `double`... Is my understanding right?

I think you could make use of inline code formatting to distinguish between `data types` and general mentions.

Comment (Calring) [reply docs/LoongArch-ELF-ABI-EN.adoc:760]: Okay, I just thought of it...

**Review by ChenghuaXu — APPROVED** [review]
